contrib/src/databases.rs

```diff
@@ -1,36 +1,116 @@
+// TODO: Look into making the connection pool type an associated type
+pub extern crate r2d2;
+
+#[cfg(feature = "diesel_sqlite")]
+pub extern crate diesel;
+
+#[cfg(feature = "pg_pool")]
+pub extern crate postgres;
+#[cfg(feature = "pg_pool")]
+pub extern crate r2d2_postgres;
+
+use std::any::Any;
+use std::collections::HashMap;
+
 use rocket::Rocket;
 use rocket::fairing::{Fairing, Info, Kind};
-use rocket::config::{Config, Table, Value};
-use rocket::config::Value::String;
+use rocket::config::Value;
 
-pub struct Databases;
+pub use self::r2d2::*;
 
 #[cfg(feature = "diesel_sqlite")]
-use diesel::SqliteConnection;
-use diesel::r2d2::ConnectionManager;
+use self::diesel::SqliteConnection;
+#[cfg(feature = "diesel_sqlite")]
+use self::diesel::r2d2::ConnectionManager;
 
 #[cfg(feature = "pg_pool")]
-use r2d2_postgres::{PostgresConnectionManager, TlsMode};
+use self::r2d2_postgres::{PostgresConnectionManager, TlsMode};
 
-#[cfg(any(feature = "diesel_sqlite", feature = "pg_pool"))]
-use r2d2::Pool;
+pub trait AdapterInitalizer {
+    type Connection: ManageConnection;
+    fn init_pool<'a>(&self, creds: &'a str, pool_size: u32) -> Pool<Self::Connection>;
+}
 
-// TODO: Assemble methods to instantiate connection pool based on configured options
-impl Databases {
-    pub fn fairing() -> Self {
-        Databases
-    }
+#[cfg(feature = "diesel_sqlite")]
+struct DieselSqliteAdapter;
+
+#[cfg(feature = "diesel_sqlite")]
+impl AdapterInitalizer for DieselSqliteAdapter {
+    type Connection = ConnectionManager<SqliteConnection>;
 
-    #[cfg(feature = "diesel_sqlite")]
-    pub fn init_sqlite_pool<'a>(&self, credentials: &'a str) -> Pool<ConnectionManager<SqliteConnection>> {
+    fn init_pool<'a>(&self, credentials: &'a str, pool_size: u32) -> Pool<Self::Connection> {
         let manager = ConnectionManager::<SqliteConnection>::new(credentials);
-        Pool::new(manager).expect("Unable to create sqlite pool")
+        Pool::builder().max_size(pool_size).build(manager).expect("Unable to create Sqlite pool")
     }
+}
+
+#[cfg(feature = "pg_pool")]
+struct PostgresAdapter;
+
+#[cfg(feature = "pg_pool")]
+impl AdapterInitalizer for PostgresAdapter {
+    type Connection = PostgresConnectionManager;
 
-    #[cfg(feature = "pg_pool")]
-    pub fn init_rust_pg_pool<'a>(&self, credentials: &'a str) -> Pool<PostgresConnectionManager> {
+    fn init_pool<'a>(&self, credentials: &'a str, pool_size: u32) -> Pool<Self::Connection> {
         let manager = PostgresConnectionManager::new(credentials, TlsMode::None).unwrap();
-        Pool::new(manager).expect("Unable to create postgres pool")
+        Pool::builder().max_size(pool_size).build(manager).expect("Unable to build Postgres pool")
+    }
+}
+
+pub struct BoxedPool(Box<Any>);
```

Comment:
This `Box<Any>` is what makes this non-Send and non-Sync and introduce the complexities with Rocket's Send + Sync requirements for things like `State<T>` and Fairings.

Reply:
Why not make this Box<Any + Sync + Send>?